pkg/wireguard/agent/agent.go

```diff
@@ -165,8 +165,12 @@ func (a *Agent) Init(ipcache *ipcache.IPCache) error {
 	defer func() {
 		// IPCache will call back into OnIPIdentityCacheChange which requires
 		// us to release a.mutex before we can add ourself as a listener.
+		//
+		// We then register to IPCache events:
+		// 1. when the fallback flag WireguardTrackAllIPsFallback is provided;
+		// 2. in native routing mode, as for tunneling we only need nodeIPs from UpdatePeer.
 		a.Unlock()
-		if addIPCacheListener {
+		if addIPCacheListener && (option.Config.WireguardTrackAllIPsFallback || !option.Config.TunnelingEnabled()) {
```

Comment:
nit: Consider adding this `(option.Config.WireguardTrackAllIPsFallback || !option.Config.TunnelingEnabled())` condition into a `bool` in the `wireguard.Agent` struct that is initialized `NewAgent`.

That way, you don't have to repeat the condition everywhere and you also don't have to do the "set + unset in defer" trick in the unit tests either.

Long-term, we probably want to do that for all reads of `option.Config`, but doing it for this long-ish condition would be a good start.

Reply:
Agree, I just addressed this comment.
I'm a bit tempted to avoid setting `a.ipCache = ipcache` in agent.Init when `needIPCacheEvents=false`, but since it is just a pointer it might not be that much beneficial not-setting it.